presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroupManager.java

```diff
@@ -0,0 +1,265 @@
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+@ThreadSafe
+public final class InternalResourceGroupManager
+        implements ResourceGroupManager
+{
+    private static final Logger log = Logger.get(InternalResourceGroupManager.class);
+    private static final File RESOURCE_GROUPS_CONFIGURATION = new File("etc/resource-groups.properties");
+    private static final String CONFIGURATION_MANAGER_PROPERTY_NAME = "resource-groups.configuration-manager";
+
+    private final ScheduledExecutorService refreshExecutor = newSingleThreadScheduledExecutor(daemonThreadsNamed("ResourceGroupManager"));
+    private final List<RootInternalResourceGroup> rootGroups = new CopyOnWriteArrayList<>();
+    private final ConcurrentMap<ResourceGroupId, InternalResourceGroup> groups = new ConcurrentHashMap<>();
+    private final AtomicReference<ResourceGroupConfigurationManager> configurationManager = new AtomicReference<>();
+    private final ClusterMemoryPoolManager memoryPoolManager;
+    private final MBeanExporter exporter;
+    private final AtomicBoolean started = new AtomicBoolean();
+    private final AtomicLong lastCpuQuotaGenerationNanos = new AtomicLong(System.nanoTime());
+    private final Map<String, ResourceGroupConfigurationManagerFactory> configurationManagerFactories = new HashMap<>();
+
+    @Inject
+    public InternalResourceGroupManager(BuiltinResourceGroupConfigurationManagerFactory builtinFactory, ClusterMemoryPoolManager memoryPoolManager, MBeanExporter exporter)
+    {
+        this.exporter = requireNonNull(exporter, "exporter is null");
+        this.memoryPoolManager = requireNonNull(memoryPoolManager, "memoryPoolManager is null");
+        requireNonNull(builtinFactory, "builtinFactory is null");
+        addConfigurationManagerFactory(builtinFactory);
+    }
+
+    @Override
+    public ResourceGroupInfo getResourceGroupInfo(ResourceGroupId id)
+    {
+        checkArgument(groups.containsKey(id), "Group %s does not exist", id);
+        return groups.get(id).getInfo();
+    }
+
+    @Override
+    public void submit(Statement statement, QueryExecution queryExecution, Executor executor)
+    {
+        ResourceGroupId group;
+        try {
+            group = selectGroup(queryExecution.getSession());
+        }
+        catch (PrestoException e) {
+            queryExecution.fail(e);
+            return;
+        }
+        createGroupIfNecessary(group, queryExecution.getSession(), executor);
+        groups.get(group).run(queryExecution);
+    }
+
+    @Override
+    public void addConfigurationManagerFactory(ResourceGroupConfigurationManagerFactory factory)
+    {
+        if (configurationManagerFactories.putIfAbsent(factory.getName(), factory) != null) {
```

Comment:
It's only called from one thread, so shouldn't matter. You're right though it should be concurrent since this class is marked thread safe